README.md
---

---

# Change_oc
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Build Status](https://travis-ci.org/austincunningham/change_oc.svg?branch=master)](https://travis-ci.org/austincunningham/change_oc)
[![npm version](https://badge.fury.io/js/change_oc.svg)](https://badge.fury.io/js/change_oc)
[![Known Vulnerabilities](https://snyk.io/test/github/austincunningham/change_oc/badge.svg)](https://snyk.io/test/github/austincunningham/change_oc)




Some command line tooling to download, extract and set the version of `oc` on the local system. 

Downloads the compressed oc binary to this module and extracts the various oc versions to the following locations
```
/opt/openshift/
       |_ 3.7
       |_ 3.9
       |_ 3.10
       |_ 3.11
       |_ 4.1
       |_ 4.2
       |_ 4.3
       |_ 4.4
       |_ 4.5
       |_ 4.6
```

Sets the oc binary to execute by symlinking
```
/usr/bin/oc
```


## Prerequisite 
- node 10
- npm 
- os linux

## Installation
Installation from npm
```bash
npm install -g change_oc
``` 


Installation from source clone and install the repo
``` bash
git clone https://github.com/austincunningham/change_oc.git
cd change_oc
npm install
```

## Usage
From npm run the module
```
change_oc
```

From source run the npm start script and follow the onscrean instructions
``` bash
npm start
```
### Download
![](./image/download.gif)
### Change binary
![](./image/change.gif)

## Release Steps
- Make changes
- If necessary update README.md
- Update CHANGELOG.md
- Squash commits to single commit
- Commit the changes
- Bump the version in the package.json
- Remove package-lock.json and run `npm install`
- Second commit for version bump
- Merge changes from the PR
- Pull upstream changes to master
- Publish the changes `npm publish`

CHANGELOG.md
---

---

# CHANGELOG
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Build Status](https://travis-ci.org/austincunningham/change_oc.svg?branch=master)](https://travis-ci.org/austincunningham/change_oc)
[![npm version](https://badge.fury.io/js/change_oc.svg)](https://badge.fury.io/js/change_oc)
[![Known Vulnerabilities](https://snyk.io/test/github/austincunningham/change_oc/badge.svg)](https://snyk.io/test/github/austincunningham/change_oc)

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Date format YYYY-MM-DD

## [Unreleased]
- ...
## [0.0.35] 2021-26-10
- rollback 4.7 binary support
## [0.0.34] 2021-19-10
- Add 4.7 binary support
## [0.0.33] 2021-26-05
- Fix issue 61 and 51, issues downloading 4.x tar's
## [0.0.32] 2020-12-05
- Add 4.6 binary support
## [0.0.31] 2020-10-09
- Add 4.5 binary support
## [0.0.30] 2020-10-08
- Add 4.4 binary support
## [0.0.29] 2020-02-21
- Add 4.3 binary support
## [0.0.28] 2019-11-04
- Add 4.2 binary support
## [0.0.27] 2019-07-15
- Fixed change script to show only what download and extracted as an option
## [0.0.26] 2019-07-11
- Add in missing package-lock.json
## [0.0.25] 2019-07-11
- Fix bug 41, url for downloading 4.1 was no longer working updated to a new one
## [0.0.24] 2019-07-02
- Fix bug if missing symlink file
## [0.0.23] 2019-06-15
- Feature add oc 4.1
- Update readme with release steps
## [0.0.22] 2019-04-07
- fix for https://app.snyk.io/org/austincunningham/project/22a0c913-0d50-4134-bd97-6a56842c3232
- remove package extract-tarball as was reporting security vulnerability
- add .npmignore
- remove node 6 from travis build
## [0.0.21] 2019-04-06
- fix for issue https://github.com/austincunningham/change_oc/issues/26
## [0.0.20] 2019-04-05
- updated the url for home page 
## [0.0.19] 2019-04-04
- Added a change log 
## [0.0.18] 2019-04-04
- Removed `prompt` package as wasn't being used and was reporting security vulnerability 
## [0.0.17] 2019-04-04
- Added synk package security 
## [0.0.16] 2019-03-31
- fix for issue https://github.com/austincunningham/change_oc/issues/30
## [0.0.15] 2019-03-28
- docs and webpage changes 
## [0.0.14] 2019-03-28
- fix for issue https://github.com/austincunningham/change_oc/issues/25
## [0.0.13]-[0.0.1]
- here there be dragons

```
___________________________________________________
@@@@@@@@@@@@@@@@@@@@@**^^""~~~"^@@^*@*@@**@@@@@@@@@
@@@@@@@@@@@@@*^^'"~   , - ' '; ,@@b. '  -e@@@@@@@@@
@@@@@@@@*^"~      . '     . ' ,@@@@(  e@*@@@@@@@@@@
@@@@@^~         .       .   ' @@@@@@, ~^@@@@@@@@@@@
@@@~ ,e**@@*e,  ,e**e, .    ' '@@@@@@e,  "*@@@@@'^@
@',e@@@@@@@@@@ e@@@@@@       ' '*@@@@@@    @@@'   0
@@@@@@@@@@@@@@@@@@@@@',e,     ;  ~^*^'    ;^~   ' 0
@@@@@@@@@@@@@@@^""^@@e@@@   .'           ,'   .'  @
@@@@@@@@@@@@@@'    '@@@@@ '         ,  ,e'  .    ;@
@@@@@@@@@@@@@' ,&&,  ^@*'     ,  .  i^"@e, ,e@e  @@
@@@@@@@@@@@@' ,@@@@,          ;  ,& !,,@@@e@@@@ e@@
@@@@@,~*@@*' ,@@@@@@e,   ',   e^~^@,   ~'@@@@@@,@@@
@@@@@@, ~" ,e@@@@@@@@@*e*@*  ,@e  @@""@e,,@@@@@@@@@
@@@@@@@@ee@@@@@@@@@@@@@@@" ,e@' ,e@' e@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@" ,@" ,e@@e,,@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@~ ,@@@,,0@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@,,@@@@@@@@@@@@@@@@@@@@@@@@@
"""""""""""""""""""""""""""""""""""""""""""""""""""

```


test/testIndex.js

const { assert } = require('chai');
const mock = require('mock-fs');
const { beforeEach, afterEach } = require('mocha');
const fixtures = require('./fixtures.json');


const { versions } = fixtures;

// eslint-disable-next-line no-undef
suite('Index tests suit', () => {
  // eslint-disable-next-line no-undef
  beforeEach(() => {
    mock({
      '/opt/': {
        'empty-dir': {/** empty directory */},
      },
    });
  });

  // eslint-disable-next-line no-undef
  afterEach(() => {
    mock.restore();
  });

  // eslint-disable-next-line no-undef
  test('first test', () => {
    console.log(versions);
    assert.equal(true, true);
  });
});


index.js
#!/usr/bin/env node

process.env.UV_THREADPOOL_SIZE = 128;
const colors = require('colors/safe');
const { exec } = require('child_process');
const execsync = require('child_process').execSync;
const fs = require('fs');
// eslint-disable-next-line camelcase
const download_oc = require('./src/download');

console.log(colors.yellow('======================== Check for OC Version Installed ============================='));
console.log(colors.yellow('The folowing versions of oc are installed at /opt/openshift'));

// Check to see if /usr/bin/oc exists
if (execsync('test /usr/bin/oc')) {
  execsync('sudo touch /usr/bin/oc');
}

// Check to see if /opt/openshift exists creates empty dir Installed
if (!fs.existsSync('/opt/openshift/Installed')) {
  execsync('sudo mkdir -p /opt/openshift/Installed');
}

function startprogram() {
  exec('ls /opt/openshift', (error, stdout, stderr) => {
    if (error) {
      console.log('error :', error);
    }
    if (stderr) {
      console.log('stderr', stderr);
    }
    console.log(colors.yellow(stdout));

    if (stdout) {
      execsync('sudo rm -rf /opt/openshift/Installed');
      download_oc(() => {
      });
    }
  });
}

startprogram();


src/extract.js
/* eslint-disable camelcase */


const colors = require('colors/safe');
const exec = require('child_process').execSync;
const fs = require('fs');
// eslint-disable-next-line camelcase
const download_oc = require('./download');
const change = require('./change');

const ocpath = 'openshift-origin-client-tools-v';

// eslint-disable-next-line func-names
const extract_oc = function (callback) {
  console.log(colors.cyan('================================= Extract OC ======================================='));
  console.log('download version: ', download_oc.version);
  const { version } = download_oc;
  fs.exists(`../${version}.tar.gz`, (err) => {
    if (err) {
      console.log(`file read error: ${err}`);
    } else {
      exec(`mkdir ${version}`);
      exec(`tar --warning=no-unknown-keyword -C ${version} -zxvf ${version}.tar.gz`);

      // move the files and remove the old directory
      if (!fs.existsSync(`/opt/openshift/${version}`)) {
        // as version 4x tar is not creating same directory structure
        if (version !== '4.1' && 
          version != '4.2' && 
          version != '4.3' && 
          version != '4.4'&& 
          version != '4.5' && 
          version != '4.6'
          ) 
        {
          const versionpath = `./${version}/${ocpath}${version}*/*`;
          exec(`mv ${versionpath} ./${version}`);
          exec(`${'rm -rf ./'}${version}/${ocpath}${version}*`);
        }
        // move extracted directroy
        exec(`sudo mv ./${version} /opt/openshift/${version}`);
        console.log(change);
        change();
      } else {
        console.log(`/opt/openshift/${version} exists already`);
        exec(`rm -rf ./${version}`);
        change();
      }
    }

    if (typeof callback === 'function') {
      callback();
    }
  });
};
module.exports = extract_oc;


src/download.js
/* eslint-disable camelcase */
const colors = require('colors/safe');
const request = require('request');
const fs = require('fs');
const readlineSync = require('readline-sync');
const cliAnimation = require('chalk-animation');
const extract_oc = require('./extract');
const change_oc = require('./change');

//
// get the version number
//

// eslint-disable-next-line camelcase
function download_oc() {
  let url;
  console.log(colors.blue('================================= Download OC ======================================='));
  const result = readlineSync.question(colors.blue('Do you wish to install an oc binary yes/no ? '));
  if (result === 'yes' || result === 'y') {
    const version = readlineSync.question(colors.blue('What version of oc do you wish to install ? \n - 3.7\n - 3.9\n - 3.10\n - 3.11\n - 4.1\n - 4.2\n - 4.3\n - 4.4\n - 4.5\n - 4.6\n'));
    exports.version = version;
    switch (version) {
      case '3.7':
        url = 'https://github.com/openshift/origin/releases/download/v3.7.2/openshift-origin-client-tools-v3.7.2-282e43f-linux-64bit.tar.gz';
        break;
      case '3.9':
        url = 'https://github.com/openshift/origin/releases/download/v3.9.0/openshift-origin-client-tools-v3.9.0-191fece-linux-64bit.tar.gz';
        break;
      case '3.10':
        url = 'https://github.com/openshift/origin/releases/download/v3.10.0/openshift-origin-client-tools-v3.10.0-dd10d17-linux-64bit.tar.gz';
        break;
      case '3.11':
        url = 'https://github.com/openshift/origin/releases/download/v3.11.0/openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit.tar.gz';
        break;
      case '4.1':
        url = 'https://mirror.openshift.com/pub/openshift-v4/clients/oc/4.1/linux/oc.tar.gz';
        break;
      case '4.2':
        url = 'https://mirror.openshift.com/pub/openshift-v4/clients/oc/4.2/linux/oc.tar.gz';
        break;
      case '4.3':
        url = 'https://mirror.openshift.com/pub/openshift-v4/clients/oc/4.3/linux/oc.tar.gz';
        break;
      case '4.4':
        url = 'https://mirror.openshift.com/pub/openshift-v4/clients/oc/4.4/linux/oc.tar.gz';
        break;
      case '4.5':
        url = 'https://mirror.openshift.com/pub/openshift-v4/clients/oc/4.5/linux/oc.tar.gz';
        break;
      case '4.6':
        url = 'https://mirror.openshift.com/pub/openshift-v4/clients/oc/4.6/linux/oc.tar.gz';
        break;
      default:
        url = false;
        console.log(colors.blue('No binary present'));
        break;
    }

    if (url && !(fs.existsSync(`${version}.tar.gz`))) {
      const req = request({
        uri: url,
        method: 'GET',
        timeout: 10000,
        followRedirect: true,
        maxRedirects: 10,
      }).pipe(fs.createWriteStream(`${version}.tar.gz`));
      // Add loading animation
      cliAnimation.radar('===================================================================================>');
      req.on('close', () => {
        console.log(colors.blue('Request finished writing to file'));
        extract_oc(() => {
          console.log('Extract completed');
        });
      });
    } else {
      console.log(colors.blue('Binary Already Download'));
      extract_oc(() => {
        console.log('Extract completed');
      });
    }
  } else {
    console.log(colors.blue('Skipping Binary Download'));
    change_oc(() => {
      console.log('Extract completed');
    });
  }
}

module.exports = download_oc;


src/checkFiles.js
const colors = require('colors/safe');
const { exec } = require('child_process');
const execsync = require('child_process').execSync;
const fs = require('fs');
// eslint-disable-next-line camelcase
const download_oc = require('./download');


function checkUsrBinOc() {
  // Check to see if /usr/bin/oc exists
  if (!fs.existsSync('/usr/bin/oc')) {
    execsync('sudo touch /usr/bin/oc');
  }
}

function checkOptOpenshift() {
// Check to see if /opt/openshift exists
  if (!fs.existsSync('/opt/openshift')) {
    execsync('sudo mkdir -p /opt/openshift/None_installed');
  }
}

function checkVersions() {
  exec('ls /opt/openshift', (error, stdout, stderr) => {
    if (error) {
      console.log('error :', error);
    }
    if (stderr) {
      console.log('stderr', stderr);
    }
    console.log(colors.yellow(stdout));

    if (stdout) {
      execsync('sudo rm -rf /opt/openshift/None_installed');
      download_oc(() => {
      });
    }
  });
}

function checkFiles() {
  console.log(colors.yellow('======================== Check for OC Version Installed ============================='));
  console.log(colors.yellow('The folowing versions of oc are installed at /opt/openshift'));

  checkUsrBinOc();

  checkOptOpenshift();

  checkVersions();
}


module.exports = checkFiles;


scripts/download-extract.sh
#!/bin/sh

echo "Enter version for download and extract : "
echo "- 3.7"
echo "- 3.9"
echo "- 3.10"
echo "- 3.11"
read oc_version

case "$oc_version" in

3.7) echo "Download and extract 3.7 "
     curl -L https://github.com/openshift/origin/releases/download/v3.7.2/openshift-origin-client-tools-v3.7.2-282e43f-linux-64bit.tar.gz --output 3.7.tar.gz
     mkdir -p ~/oc/3.7
     tar xvzf 3.7.tar.gz -C ~/oc/3.7
     mv ~/oc/3.7/open*/* ~/oc/3.7
     rm -rf ~/oc/3.7/openshift*
     ;;
3.9) echo "Download and extract 3.9 "
     curl -L https://github.com/openshift/origin/releases/download/v3.9.0/openshift-origin-client-tools-v3.9.0-191fece-linux-64bit.tar.gz --output 3.9.tar.gz
     mkdir -p ~/oc/3.9
     tar xvzf 3.9.tar.gz -C ~/Download/oc/3.9
     mv ~/oc/3.9/open*/* ~/oc/3.9
     rm -rf ~/oc/3.9/openshift*
     ;;
3.10) echo "Download and extract 3.10 "
     curl -L https://github.com/openshift/origin/releases/download/v3.10.0/openshift-origin-client-tools-v3.10.0-dd10d17-linux-64bit.tar.gz --output 3.10.tar.gz
     mkdir -p ~/oc/3.10
     tar xvzf 3.10.tar.gz -C ~/Download/oc/3.10
     mv ~/oc/3.10/open*/* ~/oc/3.10
     rm -rf ~/oc/3.10/openshift*
     ;;
3.11) echo "Download and extract 3.11 "
     curl -L https://github.com/openshift/origin/releases/download/v3.11.0/openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit.tar.gz --output 3.11.tar.gz
     mkdir -p ~/oc/3.11
     tar xvzf 3.11.tar.gz -C ~/Download/oc/3.11
     mv ~/oc/3.11/open*/* ~/oc/3.11
     rm -rf ~/oc/3.11/openshift*
     ;;
*)   echo "No such version "
     ;;
esac

     

src/change.js
/* eslint-disable max-len */

// 'use strict';
const colors = require('colors/safe');
const util = require('util');
const exec = util.promisify(require('child_process').exec);
const readlineSync = require('readline-sync');
const fs = require('fs');


// eslint-disable-next-line func-names
function change(callback) {
  console.log(colors.green('================================= Change OC ========================================'));
  console.log(colors.green('Versions Present:'));
  exec('ls /opt/openshift', (stderr, stdout) => {
    if (stderr) {
      console.log(colors.green(stderr));
    } else if (stdout) {
      console.log(colors.green(stdout));
      const version = readlineSync.question(colors.green('What version of oc do you wish to switch to ?'));
      // eslint-disable-next-line eqeqeq
      if (
         version != 3.7 && 
         version != 3.9 &&
         version != 3.10 &&
         version != 3.11 && 
         version != 4.1 && 
         version != 4.2 && 
         version != 4.3 && 
         version != 4.4 && 
         version != 4.5 && 
         version != 4.6
         ) 
        {
        console.log('Version not present');
      } else if (fs.existsSync(`/opt/openshift/${version}`)) {
        console.log('Command-line input received:');
        console.log(`Change to Version : ${version}`);
        exec(`sudo rm /usr/bin/oc && sudo ln -s /opt/openshift/${version}/oc /usr/bin/oc`)
          .then(() => exec('oc version'))
          // eslint-disable-next-line consistent-return
          .then((newVersion) => {
            console.log('New Version Number : ', newVersion.stdout);
            if (newVersion.stderr) {
              return newVersion.stderr;
            }
            if (typeof callback === 'function') {
              callback();
            }
          })
          .catch(Error);
      } else {
        console.log(`Binary version ${version} not extracted`);
      }
    }
  });
}

module.exports = change;
